From 74e0acc541ba413a2f8cfac52c9e94dab32065e3 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 20 Jan 2018 10:28:30 +0100
Subject: [PATCH 1/1] 10_linux: support loading device trees

Currently GRUB only loads the Linux kernel and the initial file system
before booting.

On the arm and arm64 architectures the device tree is also needed for
booting. The device tree is Linux kernel version dependent. So we can
not rely on the initial bootloader to provide the correct device tree
when calling grub.

Luckily the device tree to be loaded can be specified in GRUB with the
devicetree command.

The patch changes the 10_linux script to look in sequence for files
${dirname}/dtb-${version} and ${dirname}/dtb. If one of the files is
present, an entry

	devicetreee ${rel_dirname}/${dtb}

is added to the Linux boot entry.

== Distro suppport ==

Debian maintains a symbolic link /boot/dtb to the most current device
tree and another symbolic link /boot/dtb-${version} for the version
specific devicetree if package flash-kernel is installed.

On Suse a symbolic link /boot/dtb points to the current
/boot/dtb-${version}.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 util/grub.d/10_linux.in | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
index e8b01c0d0..dc3ddbe75 100644
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -143,6 +143,17 @@ linux_entry ()
 	echo	'$(echo "$message" | grub_quote)'
 	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args}
 EOF
+  if test -n "${dtb}" ; then
+    if [ x"$quiet_boot" = x0 ] || [ x"$type" != xsimple ]; then
+      message="$(gettext_printf "Loading device tree ...")"
+      sed "s/^/$submenu_indentation/" << EOF
+	echo	'$(echo "$message" | grub_quote)'
+EOF
+    fi
+    sed "s/^/$submenu_indentation/" << EOF
+	devicetree	${rel_dirname}/${dtb}
+EOF
+  fi
   if test -n "${initrd}" ; then
     # TRANSLATORS: ramdisk isn't identifier. Should be translated.
     message="$(gettext_printf "Loading initial ramdisk ...")"
@@ -244,6 +255,18 @@ while [ "x$list" != "x" ] ; do
     fi
   done
 
+  dtb=
+  for i in "dtb-${version}" "dtb" ; do
+    if test -e "${dirname}/${i}" ; then
+      dtb="${i}"
+      break
+    fi
+  done
+
+  if test -n "${dtb}" ; then
+    gettext_printf "Found dtb: %s\n" "${dirname}/${dtb}" >&2
+  fi
+
   initramfs=
   if test -n "${config}" ; then
       initramfs=`grep CONFIG_INITRAMFS_SOURCE= "${config}" | cut -f2 -d= | tr -d \"`
-- 
2.30.0

